File: redblue/redblue.py
#!/usr/bin/env python

def patternmatches(pattern, s):
    patterncount = {}
    for c in pattern:
        if c not in patterncount:
            patterncount[c] = 1
        else:
            break
    else:
        # If we didn't break, no repeating chars in pattern, so must be true.
        return len(pattern) <= len(s)
    return checkstring(s, pattern, {})

def checkstring(s, pattern, matches):
    # print("Checking string {!r}, pattern {!r}, matches {!r}".format(s, pattern, matches))
    s_len = len(s)
    if len(pattern) == 0:
        return s_len == 0
    p = pattern[0] # Pattern char to check.
    if s_len == 0 or s_len < len(pattern):
        return False # Impossible for pattern to match.

    if p in matches: # We've already seen this.
        matches_p = matches[p]
        if matches_p == s[:len(matches_p)]:
            return checkstring(s[len(matches_p):], pattern[1:], matches.copy())
        else:
            return False
    else: # p could be any number of s chars, check all possibilities:
        for i in range(1, len(s)+1):
            newmatches = matches.copy()
            newmatches[p] = s[:i]
            if checkstring(s[i:], pattern[1:], newmatches):
                vals = [v for v in newmatches.values()]
                if len(vals) == len(set(vals)):
                    return True
        else: # None of the possibilities matched.
            return False

def test(pattern, string, expect):
    try:
        assert(patternmatches(pattern, string) == expect)
    except AssertionError as e:
        print(f"Failed: {pattern!r} matches {string!r} was not {expect}")
        raise e

test("abdc", "odsihpoyywepqriohweoyafpsdoyh", True);
test("abba", "redredredred", False);
test("abba", "redbluebluered", True);
test("abba", "redbluebluereda", False);
test("abba", "abcxyzxyzabc", True);
test("abba", "abcxyzxyzabc", True);
test("baab", "abcxyzxyzabc", True);
test("dzzd", "abcxyzxyzabc", True);
test("dzzd", "dzzda", False);

test("abba", "rblblr", True)
test("abab", "redblueredblue", True)
test("abba", "catdogdogcat", True)
test("abab", "redblueredblue", True)
test("abab", "catdogcatdog", True)
test("aba","catdogcat", True)
test("abba","catdogdogcat", True)
test("abcac","catdogmousecatmouse", True)
test("abcde","efghi", True)
test("a","efghi", True)
test("abab","catdogcatcat", False)
test("abab","catdogcatdogg", False)
test("abab","catdocatdog", False)
test("abab","catdogcat", False)
test("abcdefghi","cat", False)
test("abba","redblueredblue", False)
test("aba", "patrpatrr", False)

File: redblue/redblue.js
#!/usr/bin/env node

const assert = require('assert');

function matches(pattern, input) {
  const map = {};
  let repeated_chars = false;
  for (let p of pattern) {
    if (map[p] === undefined) {
      map[p] = '';
    } else {
      repeated_chars = true;
    }
  }
  if (! repeated_chars) { return true; }
  return check_string(pattern, input, map);
}

function check_string(pattern, input, map) {
  if (pattern === '' || input === '') {
    return (pattern === '' && input === '');
  }
  let p = pattern[0];
  let p_string = map[p];
  if (p_string === '') {
    for (let i = 1; i < input.length; i++) {
      let new_map = Object.assign({}, map);
      new_map[p] = input.slice(0,i);
      if (check_string(pattern.slice(1), input.slice(i), new_map)) {
        const vals = Object.values(new_map);
        const hasDups = vals.some((value, index) => vals.indexOf(value, index + 1) !== -1);
        if (!hasDups) {
          return true;
        }
      }
    }
    return false;
  } else {
    if (p_string.length <= input.length && p_string === input.slice(0,p_string.length)) {
      return check_string(pattern.slice(1), input.slice(p_string.length), Object.assign({}, map));
    } else {
      return false;
    }
  }
}

assert.strictEqual(matches("abdc", "odsihpoyywepqriohweoyafpsdoyh"), true);
assert.strictEqual(matches("abba", "redbluebluered"), true);
assert.strictEqual(matches("abba", "redredredred"), false);
assert.strictEqual(matches("abba", "redbluebluereda"), false);
assert.strictEqual(matches("abba", "abcxyzxyzabc"), true);
assert.strictEqual(matches("abba", "abcxyzxyzabc"), true);
assert.strictEqual(matches("baab", "abcxyzxyzabc"), true);
assert.strictEqual(matches("dzzd", "abcxyzxyzabc"), true);
assert.strictEqual(matches("dzzd", "dzzda"), false);
assert.strictEqual(matches("aba", "patrpatrr"), false);

console.log('All done!');
